File: lua/exercises.lua
function change(amount)
  if math.type(amount) ~= "integer" then
    error("Amount must be an integer")
  end
  if amount < 0 then
    error("Amount cannot be negative")
  end
  local counts, remaining = {}, amount
  for _, denomination in ipairs({25, 10, 5, 1}) do
    counts[denomination] = remaining // denomination
    remaining = remaining % denomination
  end
  return counts
end

-- Write your first then lower case function here
function first_then_lower_case(sequence, predicate)
  for _, value in ipairs(sequence) do
    if predicate(value) then
      return string.lower(value)
    end
  end
  return nil
end

-- Write your powers generator here
function powers_generator(base, limit)
  return coroutine.create(function()
    local power = 1
    while power <= limit do
      coroutine.yield(power)
      power = power * base
    end
  end)
end

-- Write your say function here
function say(word)
  local words = {}

  local function chain(nextWord)
    if nextWord == nil then
      return table.concat(words, ' ')
    else
      table.insert(words, nextWord)
      return chain
    end
  end

  if word ~= nil then
    return chain(word)
  end

  return chain(nil)
end


-- Write your line count function here
function meaningful_line_count(filename)
  local count = 0
  local file, err = io.open(filename, "r")
  
  if not file then
    error("No such file")
  end
  
  local status, read_error = pcall(function()
    for line in file:lines() do
      local trimmed_line = line:match("^%s*(.-)%s*$")
      if trimmed_line and #trimmed_line > 0 and not trimmed_line:find("^#") then
        count = count + 1
      end
    end
  end)

  file:close()
  
  if not status then
    error(read_error)
  end
  
  return count
end

-- Write your Quaternion table here

Quaternion = {}
Quaternion.__index = Quaternion

function Quaternion.new(a, b, c, d)
  local self = setmetatable({}, Quaternion)
  self.a = a
  self.b = b
  self.c = c
  self.d = d
  return self
end

function Quaternion.__add(q1, q2)
  return Quaternion.new(
    q1.a + q2.a,
    q1.b + q2.b,
    q1.c + q2.c,
    q1.d + q2.d
  )
end

function Quaternion.__mul(q1, q2)
  return Quaternion.new(
    q1.a * q2.a - q1.b * q2.b - q1.c * q2.c - q1.d * q2.d,
    q1.a * q2.b + q1.b * q2.a + q1.c * q2.d - q1.d * q2.c,
    q1.a * q2.c - q1.b * q2.d + q1.c * q2.a + q1.d * q2.b,
    q1.a * q2.d + q1.b * q2.c - q1.c * q2.b + q1.d * q2.a
  )
end

function Quaternion.__eq(q1, q2)
  return q1.a == q2.a and q1.b == q2.b and q1.c == q2.c and q1.d == q2.d
end

function Quaternion:conjugate()
  return Quaternion.new(self.a, -self.b, -self.c, -self.d)
end

function Quaternion:coefficients()
  return {self.a, self.b, self.c, self.d}
end

--Couldnt figure out how to get this method to work correctly
function Quaternion:tostring()
  local terms = {}

  if self.a ~= 0 or (#terms == 0) then
    table.insert(terms, string.format("%g", self.a))
  end

  if self.b ~= 0 then
    if self.b > 0 and #terms > 0 then
      table.insert(terms, string.format("+%gi", self.b))
    else
      table.insert(terms, string.format("%gi", self.b))
    end
  end

  if self.c ~= 0 then
    if self.c > 0 and #terms > 0 then
      table.insert(terms, string.format("+%gj", self.c))
    else
      table.insert(terms, string.format("%gj", self.c))
    end
  end

  if self.d ~= 0 then
    if self.d > 0 and #terms > 0 then
      table.insert(terms, string.format("+%gk", self.d))
    else
      table.insert(terms, string.format("%gk", self.d))
    end
  end

  if #terms == 0 then
    return "0"
  end

  return table.concat(terms)
end

File: python/exercises.py
from dataclasses import dataclass
from collections.abc import Callable


def change(amount: int) -> dict[int, int]:
    if not isinstance(amount, int):
        raise TypeError('Amount must be an integer')
    if amount < 0:
        raise ValueError('Amount cannot be negative')
    counts, remaining = {}, amount
    for denomination in (25, 10, 5, 1):
        counts[denomination], remaining = divmod(remaining, denomination)
    return counts


# Write your first then lower case function here

def first_then_lower_case(a, p):
    for s in a:
        if p(s):
            return s.lower()
    return None


# Write your powers generator here

def powers_generator(base, limit):
    power = 1 
    while power <= limit:
        yield power
        power *= base


# Write your say function here

def say(word=None, accumulated=""):
    if word is None:
        return accumulated.strip()
    
    if accumulated:
        if word:
            accumulated += " " + word
        else:
            accumulated += " "
    else:
        accumulated = word

    return lambda next_word=None: say(next_word, accumulated)


# Write your line count function here

def meaningful_line_count(filename):
    try:
        file = open(filename, 'r')
        count = 0
        for line in file:
            line = line.strip()
            if line and not line.startswith('#'):
                count += 1

        file.close() 
        return count
    
    except FileNotFoundError:
        raise FileNotFoundError("No such file") 



# Write your Quaternion class here
@dataclass(frozen=True)
class Quaternion:
    a: float
    b: float
    c: float
    d: float

    @property
    def coefficients(self):
        return (self.a, self.b, self.c, self.d)

    @property
    def conjugate(self):
        return Quaternion(self.a, -self.b, -self.c, -self.d)

    def __add__(self, other):
        if isinstance(other, Quaternion):
            return Quaternion(
                self.a + other.a,
                self.b + other.b,
                self.c + other.c,
                self.d + other.d
            )
        return NotImplemented

    def __mul__(self, other):
        if isinstance(other, Quaternion):
            a = self.a * other.a - self.b * other.b - self.c * other.c - self.d * other.d
            b = self.a * other.b + self.b * other.a + self.c * other.d - self.d * other.c
            c = self.a * other.c - self.b * other.d + self.c * other.a + self.d * other.b
            d = self.a * other.d + self.b * other.c - self.c * other.b + self.d * other.a
            return Quaternion(a, b, c, d)
        return NotImplemented

    def __eq__(self, other):
        if isinstance(other, Quaternion):
            return (self.a, self.b, self.c, self.d) == (other.a, other.b, other.c, other.d)
        return False

    def __str__(self):
        terms = []
        if self.a != 0 or not any([self.b, self.c, self.d]):
            terms.append(f"{self.a}")
        if self.b != 0:
            if self.b == 1:
                terms.append("i")
            elif self.b == -1:
                terms.append("-i")
            else:
                terms.append(f"{self.b}i")
        
        if self.c != 0:
            if self.c == 1:
                terms.append("j")
            elif self.c == -1:
                terms.append("-j")
            else:
                terms.append(f"{self.c}j")
        
        if self.d != 0:
            if self.d == 1:
                terms.append("k")
            elif self.d == -1:
                terms.append("-k")
            else:
                terms.append(f"{self.d}k")
        
        result = "".join(terms)
        result = result.replace('+-', '-').lstrip('+')
        return result if result else "0"